RationaleMCP/0035/SpecificationText.md

```diff
@@ -0,0 +1,82 @@
+It contains all necessary information to translate all descriptions, but no translations. The pattern `<LibraryName>` denotes the toplevel class name of the library.
+* one file for each supported language with the name `<LibraryName>.<language>.po`. This file is a copy of the associated template file, but extended with the translations in the specified language. The pattern `<language>` stands for the ISO 639-1 language code, e.g. de or en.
+
+The files consist of a header and a body. All text strings are in double quotes and encoded with UTF-8 characters. Comments start with an `#` and are continued until the end of line. Spaces outside strings are ignored and used as separators. The detailed format of these files is described in [GNU gettext](https://www.gnu.org/software/gettext/manual/gettext.pdf).
+
+The header is marked with an empty msgid entry and looks like this:
+```
+# SOME DESCRIPTIVE TITLE.
+# Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
+# This file is distributed under the same license as the PACKAGE package.
+# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
+#
+#, fuzzy
+msgid ""
+msgstr ""
+"Project-Id-Version: PACKAGE VERSION\n"
+"Report-Msgid-Bugs-To: \n"
+"POT-Creation-Date: 2019-03-15 10:52+0100\n"
+"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\n"
+"Last-Translator: FULL NAME <EMAIL@ADDRESS>\n"
+"Language-Team: LANGUAGE <LL@li.org>\n"
+"Language: \n"
+"MIME-Version: 1.0\n"
+"Content-Type: text/plain; charset=UTF-8\n"
+"Content-Transfer-Encoding: 8bit\n"
+```
+All general terms in the header should be replaced by specific information.
+
+In the body, directly following the header, for each descriptive text there is an entry of the form
+```
+#: <file name>:<line number>
+msgctxt "<context identifier>"
+msgid "<descriptive text>"
+msgstr "<translation>"
+```
+Only the keywords `msgctxt`, `msgid` and `msgstr` are used.
+
+At first there can be an optional comment describing the location (file name and line number) of the text to translate. Multiple occurences of the same string can be listed here, separated by space.
+
+Then, the `<context identifier>` behind the keyword `msgctxt` is the full name of the Modelica class (e.g. `Modelica.Blocks.Math.Sin` ) where the text appears in. Short class definitions do not appear here. Texts in such classes belong to the enclosing full class definition.
+
+After the `msgid` keyword the text string which shall be translated follows. It should contain the original string from the Modelica text representation. 
+Since in Modelica control sequences also start with a backslash and another backslash is used to use sequences literally or to hide double quotes, no change is required here. 
+But Modelica allows strings to go over more than one line, gettext does not.
+Therefore, Modelica strings must be split into consecutive strings at the line breaks, and the line breaks must be encoded with a "\n" at the end of each string, e.g.
+```
+"A
+B"
+```
+is converted to
+```
+"A\n"
+"B"
+```

```

Comment:
Adding to my comment above, I think that the kind of splitting into separate strings as proposed here would be catastrophic from a translation point of view.  Literal line breaks are often inserted into long strings for hard line wrapping of a message to fit a certain column width.  The splitting carelessly breaks down sentences into fragments that may not translate sensibly on their own, and trying to preserve the style of hard line wrapping to the same column width would be an insane task.  (That said, I think hard line wraps in message strings are a bad idea, but it is not our business to argue over such stylistic matters here.)

Reply:
As I understand: it's **not separate** strings, but one conceptual string running over several lines.
I would assume the tools for handling translations don't care whether the string is split over several lines or not - but only care about the "\n".

Reply:
This splitting of strings serves only to support existing translation tools which follow the gettext format specification which does not support strings going over multiple lines. The tools themselves will join the lines and present them to the user as one text to translate.

Reply:
OK, thanks for the explanation.  I now found the corresponding passage in the gettext documentation.

In that case, I find this another example of something that shouldn't be dictated normatively by the Modelica specification.  A non-normative comment mentioning the possibility of splitting over multiple lines should be sufficient.  After all, with tools used to edit these files, why should we even care whether the strings are split over several lines or not?

Reply:
@HansOlsson has put the line breaking stuff in non normative text. This is only to know how to generate translation files from Modelica and how to read in multiple consecutive strings from a translation file into a Modelica tool.

Reply:
To clarify I put line break->"\n" in normative text, but the actual splitting of text for gettext in the non-normative part.

Reply:
I added 'for gettext' to make that clearer.

Reply:
I believe this is now clear, and we can start the next step in the process. @henrikt-ma  agreed?

Reply:
I'm afraid that my biggest problem is insufficient understanding of the relation between the Modelica, its specification, and the gettext specification.  I still have a feeling that the proposed Modelica specification text is too much a repetition of the gettext specification, instead of putting emphasis on how the gettext specification should be applied in the Modelica context.

One detail I don't see answered at the moment is the role of the format string, and I don't understand whether such a statement should be normative or if a non-normative remark would be sufficient.

Reply:
I don't understand what the _format string_ is intended to mean.
The basic idea is to take the strings from the Modelica code and add them almost verbatim to the gettext-file (verbatim except for the line-break handling), with context given by class-name etc.

Reply:
I'm referring to section 15.3 _The Translator's View_ in the gettext specification.  The section begins with:
> The translator works exactly as in the C language case. The only difference is that when translating format strings, she has to be aware of the language’s particular syntax for positional arguments in format strings.

To me this means that it doesn't matter what the Modelica string input syntax looks like; the strings to be translated should be evaluated (removing the difference between literal newlines and `\n`), and the string data should be expressed using C string syntax (encoding all newlines as `\n`), with some choice of positional argument syntax on top of that.

The least we could do is to comment on the role of the gettext format string syntax in the context of Modelica translations.  A simple statement like this could be enough:
> No other gettext format string syntax than `no-c-format` may be specified, and the actual semantics of a string depends on the Modelica context.  For example, `%name` has special meaning in certain Modelica contexts and must then be treated similar to positional arguments in the standardized ways of expressing gettext format strings.

Reply:
Actually, the fact that we have special meaning assigned to things like `%class` in some context means that `no-c-format` should actually be required wherever such constructs may appear.  To make things simpler, I'd suggest making `no-c-format` mandatory for all strings until we see a need to allow any other format.

Reply:
We should always put the comment `# no-c-format` in the line before msgid to avoid missinterpretation by gettext tools.

Reply:
see the change above.